src/main/java/org/broadinstitute/hellbender/tools/walkers/mutect/Mutect3DatasetEngine.java

```diff
@@ -0,0 +1,284 @@
+
+            if (trainingMode) {
+                final ArrayBlockingQueue<Integer> unmatchedQueue = unmatchedCounts.get(type);
+                final boolean likelySeqError = tumorLods[n] < TLOD_THRESHOLD;
+                final boolean likelyGermline = hasNormal && normalAF > 0.2;
+
+                // extremely strict criteria because there are so many germline variants we can afford to waste a lot
+                final boolean definiteGermline = !likelySeqError && popafs[n] < COMMON_POPAF_THRESHOLD &&
+                        tumorAF > 0.35 && (!hasNormal || normalAF > 0.35);
+
+                // low AF in tumor and normal, rare in population implies artifact
+                if  (!(likelySeqError || likelyGermline) && tumorAF < 0.2 && popafs[n] > RARE_POPAF_THRESHOLD) {
+
+                    if (unmatchedQueue.size() > 0.9*CAPACITY) { // this should rarely come up
+                        labels.add(Label.IGNORE);
+                    } else {
+                        labels.add(Label.ARTIFACT);
+                        unmatchedQueue.addAll(Collections.nCopies(nonArtifactPerArtifact, tumorADs[n + 1]));
+                    }
+                } else if (definiteGermline && !unmatchedQueue.isEmpty()) {
+                    // high AF in tumor and normal, common in population implies germline, which we downsample
+                    labels.add(Label.VARIANT);
+                    altDownsampleMap.put(vc.getAlternateAllele(n), unmatchedQueue.poll());
+                } else if (tumorLods[n] > 4.0 && tumorAF < 0.3) {
+                    labels.add(Label.UNLABELED);
+                } else {
+                    labels.add(Label.IGNORE);
+                }
+            } else {
+                labels.add(Label.UNLABELED);
+            }
+        }
+
+        Utils.validate(labels.size() == numAlt, "We have not labeled every alt, or have labeled too much");
+        // we check this later allele by allele, but we can save a lot of compute here if we realize no alt allele yields training data
+        if (trainingMode && labels.stream().allMatch(label -> label == Label.IGNORE)) {
+            return;
+        }
+
+        // haplotype equivalence counts, haplotype complexity, haplotype dominance
+        final Triple<int[], int[], double[]> assemblyComplexity = AssemblyComplexity.annotate(vc, logFragmentLikelihoods);
+
+        // TODO: for now we don't really need normal reads
+        final List<List<List<Integer>>> normalReadVectorsByAllele =  FeaturizedReadSets.getReadVectors(vc, normalSamples, likelihoods, logFragmentLikelihoods, maxRefCount, maxAltCount);
+        final List<List<List<Integer>>> tumorReadVectorsByAllele =  FeaturizedReadSets.getReadVectors(vc, tumorSamples, likelihoods, logFragmentLikelihoods, maxRefCount, maxAltCount, altDownsampleMap);
+
+        // ref reads have already been downsampled by the read featurizer
```

Comment:
both ref and alt reads have been downsampled by the read featurizer?

Reply:
thanks, fixed